include/RAJA/pattern/params/reducer.hpp

```diff
@@ -101,48 +95,99 @@ namespace detail
     unsigned int * device_count = nullptr;
 #endif
 
-    using ARG_TUP_T = camp::tuple<value_type*>;
-    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&val); }
+    // These are types and parameters extracted from this struct, and given to the forall.
+    using ARG_TUP_T = camp::tuple<VType*>;
+    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&valop_m); }
 
     using ARG_LIST_T = typename ARG_TUP_T::TList;
     static constexpr size_t num_lambda_args = camp::tuple_size<ARG_TUP_T>::value ;
   };
 
-} // namespace detail
+  // Partial specialization of Reducer for ValLoc
+  // T must be a ValLoc
+  // TTOp is the template template version of Op, required by ValOp
+  // VType must be ValOp<ValLoc<>,Op>
+  template <typename Op, typename T, template <typename, typename, typename> class TTOp>
+  struct Reducer<Op, T, ValOp<ValLoc<typename T::value_type, typename T::index_type>, TTOp>> : public ForallParamBase {
+    using target_value_type = typename T::value_type; // This should be a basic data type extracted from the T=ValLoc passed in
+    using target_index_type = typename T::index_type;
+    using value_type = T; // Note that this struct operates on ValLoc
+    using op = TTOp<value_type,value_type,value_type>;
+    using VType = ValOp<ValLoc<target_value_type,target_index_type>, TTOp>;
 
-template <template <typename, typename, typename> class Op, typename T>
-auto constexpr Reduce(T *target)
-{
-  return detail::Reducer<Op<T, T, T>, T>(target);
-}
+    Reducer() = default;
 
+    // ValLoc constructor
+    // Note that the passthru variables point to the val and loc within the user defined target ValLoc
+    RAJA_HOST_DEVICE Reducer(value_type *target_in) : valop_m(VType{}), target(target_in), passthruval(&target_in->val), passthruindex(&target_in->loc) {}
```

Comment:
If target is used should it always refer to the internal valop object to keep things simple?